Adding DuckDB type mapping and InDB support

jdunkerley opened the pull request:
### Pull Request Description

- Move `get_error_mapper` from `Dialect` to the `Type_Mapping`.
- Add `Redshift_Type_Mapping` (currently a pass-through to `Postgres_Type_Mapping` but with a different `error_mapper`).
- Remove `JDBC_Connection.create` in favour of using `JDBC_Connection.from_java`.
  - More responsibility on the typed connections.
  - The Java imports are local to the typed connection, meaning it should work in dual JVM mode.
  - New helper methods to make building the connection simpler.
- Moved `get_pragma_value` into Postgres Helper as only used there.
- New `DuckDB_Type_Mapping`. First step for full InDB support.
- Removed static `set` methods from `JDBCUtils` in favour of `JDBCValueSetter`.
- Moved the `Statement_Setter` into the type mapping and made it be created with a `JDBCValueSetter`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

jdunkerley committed:
DuckDB Type Mapping.

jdunkerley committed:
More the Error Mapper from the dialect to the Type Mapper.
(Redshift and DuckDB to fix)

jdunkerley committed:
Sort DuckDB and Redshift as well.
Adds a new Redshift_Type_Mapper.

jdunkerley committed:
Fix Redshift names and import.

jdunkerley committed:
Localise all the Java calls to make JVM split easier.

jdunkerley committed:
Docs and fixes

JaroslavTulach commented on distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Redshift_Details.enso:
```diff
@@ -64,7 +69,12 @@ type Redshift_Details
         properties = options.merge self.jdbc_properties
-        jdbc_connection = JDBC_Connection.create self.jdbc_url properties
+
+        sql_error_handler = JDBC_Connection.create_sql_error_handler SQLException SQLTimeoutException
```
Passing `SQLException` and `SQLTimeoutException` to `JDBC_Connection` will create handler capable to capture exceptions throw by this `Standard.AWS` Java code.

JaroslavTulach commented on distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Postgres_Connection.enso:
```diff
@@ -25,6 +25,13 @@ from project.Internal.Postgres.Helpers import get_encoding_name, parse_postgres_
 from project.Internal.Upload.Helpers.Default_Arguments import first_column_name_in_structure
 from project.Internal.Upload.Operations.Create import create_table_implementation
 
+polyglot java import java.sql.SQLException
```
- These copies of a block of `polyglot java import` statements are unfortunate.
- They will be hard to refactor and keep in synch
- It might be better to load just a single class that would give access to others

```java
public record EnsoSQLDriver(java.sql.Driver driver) {
  Class<SQLException sql_exception() { return SQLException.class);
  Class<Pair> pair(return Pair.class; };
}
```
by just creating instances of this class, all the other classes would be automatically loaded from the correct class loader.

JaroslavTulach commented on distribution/lib/Standard/Database/0.0.0-dev/src/Connection/SQLite_Connection.enso:
```diff
@@ -35,7 +42,11 @@ type SQLite_Connection
     create : Text -> Vector -> SQLite_Connection
     create url properties =
-        jdbc_connection = JDBC_Connection.create url properties
+        sql_error_handler = JDBC_Connection.create_sql_error_handler SQLException SQLTimeoutException
```
Another verbose copy of the same code.

JaroslavTulach reviewed (commented):
- the code in 4cfe0e4bdc568f5d54d69e19d7f05524fa3fb8dc may be correct
- but it requires a lot of copy and paste
- it might be better to have a _single class_ that would are of loading the others

jdunkerley committed:
Remove unused imports.

jdunkerley committed:
Use a record class to avoid importing everything each time.

jdunkerley committed:
Unused import

jdunkerley committed:
Docs.
Lint.
Adjust Snowflake.

AdRiley commented on distribution/lib/Standard/DuckDB/0.0.0-dev/src/Internal/DuckDB_Type_Mapping.enso:
```diff
@@ -0,0 +1,221 @@
+        Types.TIMESTAMP_WITH_TIMEZONE -> Value_Type.Date_Time True
+        Types.TIMESTAMP -> Value_Type.Date_Time False
+        Types.BLOB -> Value_Type.Binary Nothing False
+        Types.BIT -> Value_Type.Binary Nothing False
```
Would BIT not map to Boolean?

jdunkerley replied:
Not DuckDB BIT is used for BITSTRINGS

jdunkerley committed:
Fix for HideableValue - new factory method.

jdunkerley committed:
Add polyglot imports to make NI work.

jdunkerley committed:
Docs for DuckDB :(

jdunkerley commented:
Snowflake tests passed. 
A failure in cloud tests possibly due to changes and dual jvm.


@JaroslavTulach note: providing the [link to failed test run](https://github.com/enso-org/enso/actions/runs/19177421791/job/54835384886) makes easier to inspect the failure

jdunkerley committed:
Fix for failing test.

JaroslavTulach commented:
> A failure in cloud tests 

The [first exception](https://github.com/enso-org/enso/actions/runs/19177421791/job/54835384886#step:14:8612) is harmless:
- shown because logging is not really configured in the _other JVM_ and probably is too verbose, CCing @hubertp 
- it indicates that `org.sqlite.util.OSInfo.isAndroidTermux` is executing `uname` on each initialization of SQLite!
- it'd be better if our libraries did not poke around OS during their initialization

> due to changes and dual jvm.

The [other crash](https://github.com/enso-org/enso/actions/runs/19177421791/job/54835384886#step:14:8684) is real and yes, it needs to be investigated from _dual JVM_ point of view. I'm adding it to my TODO.

jdunkerley committed:
Lets see if this is enough to get through the test...

jdunkerley committed:
Slight restructure.

jdunkerley committed:
Adjust Statement_Setter to be part of Type_Mapping and create polyglot objects within correct space.

jdunkerley committed:
Docs

jdunkerley committed:
Java format.

jdunkerley committed:
Remove JDBCUtils.
Move ColumnFetcherFactory into the Type Mappings.
Alter SnowflakeJDBCUtils to be a derived JDBCValueSetter.

jdunkerley committed:
Fix

JaroslavTulach commented on std-bits/database/src/main/java/org/enso/database/JDBCValueSetter.java:
```diff
@@ -0,0 +1,68 @@
+import org.enso.base.polyglot.NumericConverter;
+import org.graalvm.polyglot.Value;
+
+public class JDBCValueSetter {
```
```suggestion
public final class JDBCValueSetter {
```

jdunkerley replied:
Snowflake has a derived version of this which handles the special case.

JaroslavTulach commented on std-bits/database/src/main/java/org/enso/database/JDBCValueSetter.java:
```diff
@@ -0,0 +1,68 @@
+    return new JDBCValueSetter(databaseName);
+  }
+
+  protected JDBCValueSetter(String databaseName) {
```
```suggestion
  private JDBCValueSetter(String databaseName) {
```

jdunkerley replied:
Needed for Snowflake.

JaroslavTulach commented on std-bits/database/src/main/java/org/enso/database/JDBCDriverTypes.java:
```diff
@@ -0,0 +1,52 @@
+import org.enso.base.enso_cloud.HideableValue;
+import org.enso.database.dryrun.OperationSynchronizer;
+
+public class JDBCDriverTypes {
```
```suggestion
public final class JDBCDriverTypes {
```

JaroslavTulach commented on std-bits/base/src/main/java/org/enso/base/enso_cloud/EnsoSecretHelper.java:
```diff
@@ -31,20 +31,20 @@ public final class EnsoSecretHelper extends SecretValueResolver {
+   * @param properties properties in the form of {@code HideableValue.KeyValuePair}
    */
   public static Connection getJDBCConnection(
-      String url, List<Pair<String, HideableValue>> properties) throws SQLException {
```
- Removing usage of pair is good idea.
- Possibly we could remove `org/graalvm/sdk/collections/25.0.0/collections-25.0.0.jar` from classpath

jdunkerley replied:
Will take a little more time as we use within HTTP requests as well and also have some use of other classes in a couple of places.

jdunkerley committed:
Fix for Snowflake.

jdunkerley committed:
Docs

jdunkerley committed:
Java format.

JaroslavTulach commented:
> The [other crash](https://github.com/enso-org/enso/actions/runs/19177421791/job/54835384886#step:14:8684) is real and yes, it needs to be investigated from _dual JVM_ point of view.

This is the Enso stack before the crash:
```
File_Connection.type.setup<arg-1> (test/Table_Tests/src/Database/SQLite_Spec.enso:404)
case_branch.case <internal-1218> (Standard/Test/0.0.0-dev/src/Group.enso:37)
Panic.type.recover (Standard/Base/0.0.0-dev/src/Panic.enso:212)
Helpers.execute_spec_code (Standard/Test/0.0.0-dev/src/Helpers.enso:55)
Runtime.no_inline (Standard/Base/0.0.0-dev/src/Runtime.enso:103)
Duration.type.time_execution (Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:150)
case_branch (Standard/Test/0.0.0-dev/src/Helpers.enso:44)
Helpers.run_spec (Standard/Test/0.0.0-dev/src/Helpers.enso:43)
case_branch.test_results (Standard/Test/0.0.0-dev/src/Helpers.enso:24)
Array_Like_Helpers.map.Array_Like_Helpers.map (Standard/Base/0.0.0-dev/src/Internal/Array_Like_Helpers.enso:277)
case_branch (Standard/Test/0.0.0-dev/src/Helpers.enso:21)
Helpers.run_specs_from_group (Standard/Test/0.0.0-dev/src/Helpers.enso:18)
case_branch (Standard/Test/0.0.0-dev/src/Suite.enso:101)
Suite.run_with_filter.junit_sb_builder (Standard/Test/0.0.0-dev/src/Suite.enso:99)
Array_Like_Helpers.each.Array_Like_Helpers.each (Standard/Base/0.0.0-dev/src/Internal/Array_Like_Helpers.enso:317)
Range.each.go (Standard/Base/0.0.0-dev/src/Data/Range.enso:322)
Test_Reporter.wrap_junit_testsuites (Standard/Test/0.0.0-dev/src/Test_Reporter.enso:30)
Suite.run_with_filter (Standard/Test/0.0.0-dev/src/Suite.enso:95)
SQLite_Spec.main (test/Table_Tests/src/Database/SQLite_Spec.enso:482)
```
the crash then continues by calling into `org.sqlite.jdbc4.JDBC4Connection.close()` which then calls native method `_close()`  . This is the change I use:
```diff
diff --git test/Table_Tests/src/Database/SQLite_Spec.enso test/Table_Tests/src/Database/SQLite_Spec.enso
index ee4339ba43..51fac18d4e 100644
--- test/Table_Tests/src/Database/SQLite_Spec.enso
+++ test/Table_Tests/src/Database/SQLite_Spec.enso
@@ -4,6 +4,7 @@ from Standard.Base.Runtime import assert
 import Standard.Base.Errors.File_Error.File_Error
 import Standard.Base.Errors.Illegal_Argument.Illegal_Argument
 import Standard.Base.Runtime.Context
+import Standard.Base.Runtime.Debug
 
 from Standard.Table import Table, Value_Type
 from Standard.Table.Errors import Invalid_Column_Names, Duplicate_Output_Column_Names
@@ -398,10 +399,16 @@ type File_Connection
     Value ~file
 
     setup db_file:Database_File = File_Connection.Value <|
+        IO.println "setup 1"
         tmp_file = db_file.file
+        IO.println "setup 2"
         con = create_file_connection tmp_file
+        IO.println "setup 3:"+con.to_text+"   "+(Meta.type_of con).to_text
+        IO.readln
         con.close
+        IO.println "setup 4"
         assert tmp_file.exists
+        IO.println "setup 5"
         tmp_file
 
 
@@ -412,12 +419,12 @@ add_specs suite_builder =
     # https://github.com/enso-org/enso/issues/9437
     database_file = Database_File.create
 
-    sqlite_spec suite_builder in_file_prefix (_ -> create_file_connection database_file.file) persistent_connector=True
-    Transaction_Spec.add_specs suite_builder (_ -> create_file_connection database_file.file) in_file_prefix
+    #sqlite_spec suite_builder in_file_prefix (_ -> create_file_connection database_file.file) persistent_connector=True
+    #Transaction_Spec.add_specs suite_builder (_ -> create_file_connection database_file.file) in_file_prefix
 
-    in_memory_prefix = "[SQLite In-Memory] "
-    sqlite_spec suite_builder in_memory_prefix (_ -> create_inmem_connection) persistent_connector=False
-    Transaction_Spec.add_specs suite_builder (_ -> create_inmem_connection) in_memory_prefix
+    #in_memory_prefix = "[SQLite In-Memory] "
+    #sqlite_spec suite_builder in_memory_prefix (_ -> create_inmem_connection) persistent_connector=False
+    #Transaction_Spec.add_specs suite_builder (_ -> create_inmem_connection) in_memory_prefix
 
     SQLite_Type_Mapping_Spec.add_specs suite_builder
     Generic_JDBC_SQLite_Spec.add_specs suite_builder
```
the program stops and waits for Enter. Then it crashes.

JaroslavTulach commented:
Heuréka. I finally got info from `gdb`:
```
Thread 1 "enso" received signal SIGSEGV, Segmentation fault.
0x00007ff7dae0ee75 in ?? () from /home/devel/NetBeansProjects/enso/enso.ide/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Database/0.0.0-dev/polyglot/lib/amd64/linux/libsqlitejdbc.so
(gdb) bt
#0  0x00007ff7dae0ee75 in ?? () from /home/devel/NetBeansProjects/enso/enso.ide/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Database/0.0.0-dev/polyglot/lib/amd64/linux/libsqlitejdbc.so
#1  0x00007ff7dae0df87 in ?? () from /home/devel/NetBeansProjects/enso/enso.ide/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Database/0.0.0-dev/polyglot/lib/amd64/linux/libsqlitejdbc.so
#2  0x00007ff7dae0fb51 in Java_org_sqlite_core_NativeDB__1close ()
   from /home/devel/NetBeansProjects/enso/enso.ide/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Database/0.0.0-dev/polyglot/lib/amd64/linux/libsqlitejdbc.so
#3  0x000055555b8280f9 in org.sqlite.core.NativeDB::_close() (this=<optimized out>) at org/sqlite/core/NativeDB.java:1
#4  0x000055555b823f70 in org.sqlite.core.DB::close() (this=<optimized out>) at org/sqlite/core/DB.java:246
#5  0x000055555b814e96 in org.sqlite.SQLiteConnection::close() (this=<optimized out>) at org/sqlite/SQLiteConnection.java:420
#6  0x0000555557396acb in com.oracle.svm.core.reflect.ReflectionAccessorHolder::invoke_55THlcps1uCAiZuordqMDE(java.lang.Object*, java.lang.Object[]*, org.graalvm.nativeimage.c.function.CFunctionPointer) (__0=<optimized out>, 
    __1=<optimized out>, __2=<optimized out>) at com/oracle/svm/core/reflect/ReflectionAccessorHolder.java:1
#7  0x00005555575105aa in com.oracle.svm.core.reflect.SubstrateMethodAccessor::invoke(java.lang.Object*, java.lang.Object[]*) (this=<optimized out>, obj=<optimized out>, args=<optimized out>)
    at com/oracle/svm/core/reflect/SubstrateMethodAccessor.java:122
#8  0x00005555578a748d in java.lang.reflect.Method::invoke(java.lang.Object*, java.lang.Object[]*) (this=<optimized out>, obj=<optimized out>, args=<optimized out>) at java/lang/reflect/Method.java:565
#9  com.oracle.truffle.host.HostMethodDesc$SingleMethod$MethodReflectImpl::reflectInvoke(java.lang.reflect.Method*, java.lang.Object*, java.lang.Object[]*) (reflectionMethod=<optimized out>, receiver=<optimized out>, 
    arguments=<optimized out>) at com/oracle/truffle/host/HostMethodDesc.java:306
#10 0x00005555578a71e5 in com.oracle.truffle.host.HostMethodDesc$SingleMethod$MethodReflectImpl::invoke(java.lang.Object*, java.lang.Object[]*) (this=0x7ff805007d68, receiver=<optimized out>, arguments=<optimized out>)
    at com/oracle/truffle/host/HostMethodDesc.java:291
#11 0x0000555557870f31 in com.oracle.truffle.host.GuestToHostCodeCache$GuestToHostInvokeReflect::executeImpl(java.lang.Object*, java.lang.Object[]*) (this=<optimized out>, obj=<optimized out>, callArguments=<optimized out>)
    at com/oracle/truffle/host/GuestToHostCodeCache.java:110
#12 0x0000555557873909 in com.oracle.truffle.host.GuestToHostRootNode::execute(com.oracle.truffle.api.frame.VirtualFrame*) (this=<optimized out>, frame=<optimized out>) at com/oracle/truffle/host/GuestToHostRootNode.java:80
#13 0x0000555557d4b8fd in com.oracle.truffle.runtime.OptimizedCallTarget::executeRootNode(com.oracle.truffle.api.frame.VirtualFrame*, com.oracle.truffle.runtime.CompilationState*) (this=0x7ff80f2ceff0, frame=0x7ff80356b470, 
    tier=0x7ff7feb6c940) at com/oracle/truffle/runtime/OptimizedCallTarget.java:830
#14 0x0000555557d49389 in com.oracle.truffle.runtime.OptimizedCallTarget::callInlined(com.oracle.truffle.api.nodes.Node*, java.lang.Object[]*) (this=<optimized out>, location=0x7ff809a6c4f8, arguments=<optimized out>)
    at com/oracle/truffle/runtime/OptimizedCallTarget.java:626
#15 0x0000555557d6e510 in com.oracle.truffle.runtime.OptimizedRuntimeSupport::callInlined(com.oracle.truffle.api.nodes.Node*, com.oracle.truffle.api.CallTarget*, java.lang.Object[]*) (this=<optimized out>, callNode=0x7ff809a6c4f8, 
    target=<optimized out>, arguments=<optimized out>) at com/oracle/truffle/runtime/OptimizedRuntimeSupport.java:292
#16 0x0000555557873c77 in com.oracle.truffle.host.GuestToHostRootNode::guestToHostCall(com.oracle.truffle.api.nodes.Node*, com.oracle.truffle.api.CallTarget*, java.lang.Object[]*) (node=<optimized out>, target=<optimized out>, 
    arguments=<optimized out>) at com/oracle/truffle/host/GuestToHostRootNode.java:102
#17 0x00005555578a7a00 in com.oracle.truffle.host.HostMethodDesc$SingleMethod$ReflectBase::invokeGuestToHost(java.lang.Object*, java.lang.Object[]*, com.oracle.truffle.host.GuestToHostCodeCache*, com.oracle.truffle.host.HostContext*, com.oracle.truffle.api.nodes.Node*) (this=<optimized out>, receiver=<optimized out>, arguments=<optimized out>, cache=<optimized out>, hostContext=<optimized out>, node=<optimized out>) at com/oracle/truffle/host/HostMethodDesc.java:260
#18 0x0000555557887013 in com.oracle.truffle.host.HostExecuteNode::doInvoke(com.oracle.truffle.api.nodes.Node*, com.oracle.truffle.host.HostMethodDesc$SingleMethod*, java.lang.Object*, java.lang.Object[]*, com.oracle.truffle.host.GuestToHostCodeCache*, com.oracle.truffle.host.HostContext*, com.oracle.truffle.host.HostContext$ToGuestValueNode*) (node=0x7ff809a6c4f8, method=<optimized out>, obj=<optimized out>, arguments=<optimized out>, cache=<optimized out>, 
    hostContext=0x7ff80f3958a8, toGuest=0x7ff7fe4d1328) at com/oracle/truffle/host/HostExecuteNode.java:906
#19 0x0000555557886994 in com.oracle.truffle.host.HostExecuteNode::doFixed(com.oracle.truffle.api.nodes.Node*, com.oracle.truffle.host.HostMethodDesc$SingleMethod*, java.lang.Object*, java.lang.Object[]*, com.oracle.truffle.host.HostContext*, com.oracle.truffle.host.HostMethodDesc$SingleMethod*, com.oracle.truffle.host.HostToTypeNode[]*, com.oracle.truffle.host.HostContext$ToGuestValueNode*, com.oracle.truffle.api.profiles.InlinedExactClassProfile*, com.oracle.truffle.api.profiles.InlinedBranchProfile*, com.oracle.truffle.api.profiles.InlinedBranchProfile*, com.oracle.truffle.host.GuestToHostCodeCache*) (node=0x7ff809a6c4f8, method=<optimized out>, obj=<optimized out>, args=<optimized out>, 
    hostContext=<optimized out>, cachedMethod=0x7ff805007d68, toJavaNodes=<optimized out>, toGuest=<optimized out>, receiverProfile=<optimized out>, errorBranch=<optimized out>, seenDynamicScope=0x7ff7ffa841b8, cache=<optimized out>)
    at com/oracle/truffle/host/HostExecuteNode.java:140
#20 0x00005555578990a8 in com.oracle.truffle.host.HostExecuteNodeGen$Inlined::execute(com.oracle.truffle.api.nodes.Node*, com.oracle.truffle.host.HostMethodDesc*, java.lang.Object*, java.lang.Object[]*, com.oracle.truffle.host.HostContext*) (this=<optimized out>, arg0Value=<optimized out>, arg1Value=<optimized out>, arg2Value=<optimized out>, arg3Value=<optimized out>, arg4Value=<optimized out>) at com/oracle/truffle/host/HostExecuteNodeGen.java:272
#21 0x00005555578cce12 in com.oracle.truffle.host.HostObject::invokeMember(java.lang.String*, java.lang.Object[]*, com.oracle.truffle.api.nodes.Node*, com.oracle.truffle.host.HostObject$LookupMethodNode*, com.oracle.truffle.host.HostExecuteNode*, com.oracle.truffle.host.HostObject$LookupFieldNode*, com.oracle.truffle.host.HostObject$ReadFieldNode*, com.oracle.truffle.api.interop.InteropLibrary*, com.oracle.truffle.api.profiles.InlinedBranchProfile*) (
    this=<optimized out>, name=<optimized out>, args=<optimized out>, node=<optimized out>, lookupMethod=<optimized out>, executeMethod=<optimized out>, lookupField=<optimized out>, readField=<optimized out>, 
--Type <RET> for more, q to quit, c to continue without paging--
    fieldValues=<optimized out>, error=<optimized out>) at com/oracle/truffle/host/HostObject.java:465
#22 0x00005555578f5605 in com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached::invokeMember(java.lang.Object*, java.lang.String*, java.lang.Object[]*) (this=<optimized out>, arg0Value_=<optimized out>, 
    arg1Value=<optimized out>, arg2Value=<optimized out>) at com/oracle/truffle/host/HostObjectGen.java:6988
#23 0x000055555768cfc8 in com.oracle.truffle.api.interop.InteropLibrary$Asserts::invokeMember(java.lang.Object*, java.lang.String*, java.lang.Object[]*) (this=0x7ff803c28de0, receiver=0x7ff8070b6c50, identifier=0x7ff7fa065978, 
    arguments=<optimized out>) at com/oracle/truffle/api/interop/InteropLibrary.java:3596
#24 0x00005555576b60f5 in com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch::invokeMember(java.lang.Object*, java.lang.String*, java.lang.Object[]*) (this=<optimized out>, receiver_=<optimized out>, 
    member=<optimized out>, arguments=<optimized out>) at com/oracle/truffle/api/interop/InteropLibraryGen.java:8497
#25 0x000055555ab1fc8f in org.enso.interpreter.node.callable.resolver.HostMethodCallNode::resolveHostMethod(org.enso.interpreter.node.callable.resolver.HostMethodCallNode$PolyglotCallType*, java.lang.String*, java.lang.Object*, java.lang.Object[]*, com.oracle.truffle.api.interop.InteropLibrary*, org.enso.interpreter.node.expression.foreign.HostValueToEnsoNode*) (this=0x7ff804636650, callType=<optimized out>, symbol=<optimized out>, self=0x7ff8070b6c50, 
    args=<optimized out>, members=<optimized out>, hostValueToEnsoNode=<optimized out>) at org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:223
#26 0x000055555ab20c65 in org.enso.interpreter.node.callable.resolver.HostMethodCallNodeGen::execute(org.enso.interpreter.node.callable.resolver.HostMethodCallNode$PolyglotCallType*, java.lang.String*, java.lang.Object*, java.lang.Object[]*) (this=<optimized out>, arg0Value=<optimized out>, arg1Value=<optimized out>, arg2Value=<optimized out>, arg3Value=<optimized out>) at org/enso/interpreter/node/callable/resolver/HostMethodCallNodeGen.java:77
#27 0x000055555aad176e in org.enso.interpreter.node.callable.InstanceInvokeMethodNode::doPolyglot(com.oracle.truffle.api.frame.VirtualFrame*, org.enso.interpreter.runtime.state.State*, org.enso.interpreter.runtime.callable.UnresolvedSymbol*, java.lang.Object*, java.lang.Object[]*, org.enso.interpreter.runtime.library.dispatch.TypesLibrary*, com.oracle.truffle.api.interop.InteropLibrary*, org.enso.interpreter.runtime.warning.WarningsLibrary*, org.enso.interpreter.node.callable.resolver.MethodResolverNode*, org.enso.interpreter.node.callable.resolver.HostMethodCallNode$PolyglotCallType*, org.enso.interpreter.node.callable.thunk.ThunkExecutorNode[]*, com.oracle.truffle.api.profiles.BranchProfile[]*, com.oracle.truffle.api.profiles.BranchProfile[]*, com.oracle.truffle.api.profiles.BranchProfile*, org.enso.interpreter.node.callable.resolver.HostMethodCallNode*, org.enso.interpreter.runtime.warning.AppendWarningNode*, org.enso.interpreter.runtime.data.hash.HashMapSizeNode*, org.enso.interpreter.runtime.data.hash.HashMapInsertAllNode*) (this=<optimized out>, frame=<optimized out>, state=<optimized out>, symbol=<optimized out>, self=<optimized out>, 
    arguments=<optimized out>, methods=<optimized out>, interop=<optimized out>, warnings=<optimized out>, preResolveMethod=<optimized out>, polyglotCallType=<optimized out>, argExecutors=<optimized out>, profiles=<optimized out>, 
    warningProfiles=<optimized out>, anyWarningsProfile=0x7ff804636638, hostMethodCallNode=<optimized out>, appendWarningNode=0x7ff804982428, mapSizeNode=<optimized out>, mapInsertAllNode=<optimized out>)
    at org/enso/interpreter/node/callable/InstanceInvokeMethodNode.java:450
#28 0x000055555aaddb67 in org.enso.interpreter.node.callable.InstanceInvokeMethodNodeGen::execute(com.oracle.truffle.api.frame.VirtualFrame*, org.enso.interpreter.runtime.state.State*, org.enso.interpreter.runtime.callable.UnresolvedSymbol*, java.lang.Object*, java.lang.Object[]*) (this=<optimized out>, frameValue=<optimized out>, arg0Value=<optimized out>, arg1Value=<optimized out>, arg2Value=<optimized out>, arg3Value=<optimized out>)
    at org/enso/interpreter/node/callable/InstanceInvokeMethodNodeGen.java:294
```
it shows that the `libsqlite.so` file is used from **both JVMs**! That's probably not a good idea. The C code is probably full of assumptions that it runs in a single JVM.

jdunkerley committed:
Move Generic JDBC test to Generic JDBC!

jdunkerley commented:
Passed all snowflake and cloud tests.

JaroslavTulach commented on test/Table_Tests/src/Database/SQLite_Spec.enso:
```diff
@@ -420,7 +419,6 @@ add_specs suite_builder =
     Transaction_Spec.add_specs suite_builder (_ -> create_inmem_connection) in_memory_prefix
 
     SQLite_Type_Mapping_Spec.add_specs suite_builder
-    Generic_JDBC_SQLite_Spec.add_specs suite_builder
```
- the 146f74f13cfc3481a89c2bef7bb8f97d56ee9b98 changes is simple
- compared to a change where we would try to use the same `.so` from multiple JVMs
- I think the change also captures the spirit of the use-case:
   - there is little point of using SQLite via generic JDBC and directly via its driver
- the SEGFAULT crash will also go away if SQLite is packaged as own library and loaded via HotSpot
